noir-projects/aztec-nr/aztec/src/keys/getters/mod.nr

```diff
@@ -0,0 +1,153 @@
+        get_historical_public_keys(self, address).ivpk_m
+    }
+
+    fn get_ovpk_m(self, _context: &mut PrivateContext, address: AztecAddress) -> Point {
+        get_historical_public_keys(self, address).ovpk_m
+    }
+
+    fn get_tpk_m(self, _context: &mut PrivateContext, address: AztecAddress) -> Point {
+        get_historical_public_keys(self, address).tpk_m
+    }
+
+    fn get_npk_m_hash(self, context: &mut PrivateContext, address: AztecAddress) -> Field {
+        self.get_npk_m(context, address).hash()
+    }
+}
+// docs:end:key-getters
+
+// A helper function since requesting nsk_app is very common
+// TODO(#6543)
+pub fn get_nsk_app(npk_m_hash: Field) -> Field {
+    get_key_validation_request(npk_m_hash, NULLIFIER_INDEX).sk_app
+}
+
+// This is the number of blocks that must pass after a key rotation event until the old keys are fully phased out and
+// become invalid.
+global KEY_REGISTRY_UPDATE_BLOCKS = 5;
+
+global KEY_REGISTRY_STORAGE_SLOT = 1;
+
+// Returns all current public keys for a given account, applying proper constraints to the context. We read all
+// keys at once since the constraints for reading them all are actually fewer than if we read them one at a time - any
+// read keys that are not required by the caller can simply be discarded.
+pub fn get_current_public_keys(context: &mut PrivateContext, account: AztecAddress) -> PublicKeys {
+    // We're going to perform historical reads from public storage, and so need to constrain the caller so that they
+    // cannot use very old blocks when constructing proofs, and hence e.g. read very old keys. We are lax and allow
+    // _any_ recent block number to be used, regardless of whether there may have been a recent key rotation. This means
+    // that multiple sets of keys are valid for a while immediately after rotation, until the old keys become phased
+    // out. We *must* be lax to prevent denial of service and transaction fingerprinting attacks by accounts that rotate
+    // their keys frequently.
+    // Note that we constrain the max block number even if the registry ends up being empty: this ensures that proof of
+    // an empty registry is also fresh.
+    let current_header = context.get_header();
+    context.set_tx_max_block_number(current_header.global_variables.block_number as u32 + KEY_REGISTRY_UPDATE_BLOCKS);
+
+    get_historical_public_keys(current_header, account)
+}
+
+// Returns historical public keys for a given account at some block determined by a block header. We read all keys at
+// once since the constraints for reading them all are actually fewer than if we read them one at a time - any read keys
+// that are not required by the caller can simply be discarded.
+// WARNING: if called with a historical header created from a fixed block this function will explictly ignores key
+// rotation! This means that callers of this may force a user to use old keys, potentially leaking privacy. Only call
```

Comment:
```suggestion
// rotation! This means that callers of this may force a user to use old keys, potentially leaking privacy (if the old keys were leaked). Only call
```

It was not immediately clear to me how using old keys would leak privacy.